### PR TITLE
[jemalloc] set correct excess in realloc_excess

### PR DESCRIPTION
**Pull request by gnzlbg**

### CODE REVIEW

File: src/liballoc_jemalloc/lib.rs [hunk 1]
```diff
@@ -211,17 +215,28 @@ mod contents {
...
+        if new_align != old_align {
+            ptr::write(err as *mut AllocErr,
+                       AllocErr::Unsupported { details: "can't change alignments" });
+            return 0 as *mut u8
         }
-        return p
+
+        let flags = align_to_flags(new_align);
+        let ptr = rallocx(ptr as *mut c_void, new_size, flags) as *mut u8;
+        let alloc_size = sallocx(ptr as *mut c_void, flags);
```
Comment (arthurprs) [inline src/liballoc_jemalloc/lib.rs]: `sallocx` and `*excess =` need to go into a !null branch.

Comment (gnzlbg) [reply src/liballoc_jemalloc/lib.rs]: !!! indeed, should be fixed now

**Comment by gnzlbg** [comment]
@kennytm can you try this?

**Comment by gnzlbg** [comment]
@Mark-Simulacrum would it be possible to run a perf of this?

**Comment by arthurprs** [comment]
Is this function used anywhere right now?

**Comment by gnzlbg** [comment]
@arthurprs no idea

---

FYI jemalloc devs say that `nallocx` should give the correct result here.

**Comment by arthurprs** [comment]
`nallocx` seems to be faster, I didn't dig into the code to find why.

```
running 3 tests
test roundtrip           ... bench:          13 ns/iter (+/- 0)
test roundtrip_w_nallocx ... bench:          15 ns/iter (+/- 0)
test roundtrip_w_sallocx ... bench:          19 ns/iter (+/- 1)



#[bench]
fn roundtrip_w_nallocx(b: &mut Bencher) {
    b.iter(|| unsafe {
        let ptr = jemalloc_sys::mallocx(100, 0);
        let rsz = jemalloc_sys::nallocx(100, 0);
        jemalloc_sys::sdallocx(ptr, rsz, 0);
        // rsz
    });
}

#[bench]
fn roundtrip_w_sallocx(b: &mut Bencher) {
    b.iter(|| unsafe {
        let ptr = jemalloc_sys::mallocx(100, 0);
        let rsz = sallocx(ptr as usize, 0);
        jemalloc_sys::sdallocx(ptr, rsz, 0);
        // rsz
    });
}

#[bench]
fn roundtrip(b: &mut Bencher) {
    b.iter(|| unsafe {
        let ptr = jemalloc_sys::mallocx(100, 0);
        // let rsz = sallocx(ptr as usize, 0);
        jemalloc_sys::sdallocx(ptr, 100, 0);
    });
}
```

**Comment by gnzlbg** [comment]
> `nallocx` seems to be faster. 

This is what I expected, IIUC `nallocx` just needs to compute a size, while `sallocx` might need to traverse some internal data-structures in some cases.

As soon as @Mark-Simulacrum does a perf run, I will commit the version with `nallocx`, so that we can compare them, and probably right after just merge the `nallocx` version. I just don't want to be walking over "blind" assumptions when working on `RawVec`.

**Comment by Mark-Simulacrum** [comment]
You can push and run try while you wait for perf; I just need the artifacts from try builds which stick around for 30 days.

**Comment by gnzlbg** [comment]
@kennytm  could you try the last commit ?

**Comment by Mark-Simulacrum** [comment]
First commit: http://perf.rust-lang.org/compare.html?start=f764eaf453162cd19ef484ece07cc21e14dfb2c1&end=4b59c8253d8080e5dbf71807b4d8fb6625760910&stat=instructions%3Au

Second commit: http://perf.rust-lang.org/compare.html?start=56dc171a2f2dc2373b1930f71140a354fa84b982&end=9b02a2d18a4c02a7781ac9ad4e76c6c16cf5961b&stat=instructions%3Au

**Comment by gnzlbg** [comment]
@bluss a similar effect to the one in  #45512 can be observed here.

The micro-benchmarks of @arthurprs show that `sallocx` is slower, but perf shows that it is a bit faster. The benches with `sallocx` use less memory and are faster. The only explanation I can think of here is that `sallocx` is maybe more accurate than `nallocx`. I am going to ping jemalloc devs on these issues.

**Comment by arthurprs** [comment]
I'm a bit confused. What are we measuring? Is this function even used in the current codebase?

**Comment by gnzlbg** [comment]
@arthurprs it is not used anywhere in `std` or `core` but I don't know if those crates use it. 

If they don't, this change should have no effect on the benchmarks but e.g. memory consumption in some cases varies up to 10-15%, that would be a lot of noise. The same applies to the other PR.

**Comment by arthurprs** [comment]
I don't think it's used, can you double check?

**Comment by gnzlbg** [comment]
@arthurprs one of the regex benchmarks shows a 15% difference. I've followed all dependencies of the `regex` crate and couldn't find any that uses `alloc_excess` or `realloc_excess`. That would mean that 19% differences in memory consumption (shown by the other PR) are just noise. 

@Mark-Simulacrum thoughts? I think `perf` runs should come with an error estimate for each test.

**Comment by arthurprs** [comment]
I think some of these tests are showing a high variance even in unrelated changes, example graph http://perf.rust-lang.org/index.html?start=2017-10-25&end=2017-10-29&stat=max-rss&absolute=false

**Comment by shepmaster** [comment]
Moving to another libs team member...

r? @sfackler

**Comment by gnzlbg** [comment]
I think this is good to go, ideally, I (or @arthurprs) will commit @arthurprs synthetic benchmarks to rustc (currently `liballoc_jemalloc` has no benchmarks), but we can do that later.

**Comment by sfackler** [comment]
I may be missing some background here, but why are we optimizing a function that isn't called anywhere?

**Comment by gnzlbg** [comment]
Because we want to be able to call it from `RawVec`. Right now `RawVec` just reallocs dropping the excess storage returned by the allocator. We can use `usable_size` to hack around this there, but... this is actually why `realloc_excess` exists in the first place.

The fact that `realloc_excess` drops the excess storage is a bug.

**Comment by sfackler** [comment]
But were the perf comparisons above performed with that RawVec change?

**Comment by gnzlbg** [comment]
@sfackler the perf comparisons above are completely irrelevant because the compiler doesn't use `realloc_excess` anywhere but they only test for the performance of the compiler. 

@arthurprs did some synthetic benchmarks: https://github.com/rust-lang/rust/pull/45514#issuecomment-339397082 where the cost of calling `nallocx` is of about ~2 ns on its machine. 

Once we enable this in `RawVec` we'll need to add synthetic benchmarks for it because the `perf` results have too much noise.

**Comment by sfackler** [comment]
Ok cool, that makes sense. Is there any reason not to land both this and the alloc_excess change or do you want to leave one open for now?

**Comment by gnzlbg** [comment]
The reason was that I wanted to run perf on both separately because... 1) I thought perf benchmarked something more than the compiler (which it doesn't; external crates could have been using `ralloc/alloc_excess`), and 2) I wanted to check how these impacted memory consumption in real programs (which perf doesn't show because it only test the compiler and the variance in memory consumption is way too high). 

Had I knew all of this this would have been merged in @arthurprs together one week ago.

**Comment by sfackler** [comment]
Ok, let's merge both into this PR and land it to avoid having to wait on bors twice.

**Comment by gnzlbg** [comment]
@sfackler done, I am closing #45512 .

**Pull request merged**